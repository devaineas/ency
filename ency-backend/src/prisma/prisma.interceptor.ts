import {
  BadRequestException,
  CallHandler,
  ExecutionContext,
  Injectable,
  NestInterceptor,
} from "@nestjs/common";
import { PrismaClientKnownRequestError } from "@prisma/client/runtime";
import { Observable, throwError } from "rxjs";
import { catchError } from "rxjs/operators";

/**
 * Handles errors of type PrismaClientKnownRequestError
 */
@Injectable()
export class PrismaInterceptor implements NestInterceptor {
  intercept(_context: ExecutionContext, next: CallHandler): Observable<any> {
    return next.handle().pipe(
      catchError((error) => {
        if (error instanceof PrismaClientKnownRequestError) {
          if (error.code === "P2002") {
            throw new BadRequestException("Already exists");
          }
        }
        return throwError(error);
      }),
    );
  }
}
