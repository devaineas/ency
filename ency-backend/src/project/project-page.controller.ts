import {
  Controller,
  Get,
  Post,
  Body,
  Put,
  Param,
  Delete,
  ParseIntPipe,
} from "@nestjs/common";
import { ProjectPageService } from "./project-page.service";
import { CreateProjectPageDto } from "./dto/create-project-page.dto";
import { UpdateProjectPageDto } from "./dto/update-project-page.dto";
import { Public } from "src/auth/auth.guard";
import { FirebaseUser } from "src/auth/decorators/user.decorator";
import { IFirebaseUser } from "src/auth/interfaces/user.interface";
import {
  ProjectPageEntity,
  ProjectPageEntityNoData,
} from "./enitiies/projectPage.entity";

/**
 * Here id param will be used to refer to the Projects ID and pageId for ProjectPages ID
 */
@Controller("project")
export class ProjectPageController {
  constructor(private readonly projectPageService: ProjectPageService) {}

  @Post(":id/page")
  create(
    @FirebaseUser() user: IFirebaseUser,
    @Body() createProjectPageDto: CreateProjectPageDto,
  ) {
    return this.projectPageService.create(user, createProjectPageDto);
  }

  @Get(":id/page")
  @Public()
  findAll(
    @FirebaseUser() user: IFirebaseUser,
    @Param("id", ParseIntPipe) projectId: number,
  ): Promise<ProjectPageEntityNoData[]> {
    return this.projectPageService.findAll(user, projectId);
  }

  @Get(":id/page/:pageId")
  @Public()
  findOne(
    @FirebaseUser() user: IFirebaseUser,
    @Param("pageId", ParseIntPipe) pageId: number,
  ): Promise<ProjectPageEntity> {
    return this.projectPageService.findOne(user, pageId);
  }

  @Put(":id/page/:pageId")
  update(
    @FirebaseUser() user: IFirebaseUser,
    @Param("pageId", ParseIntPipe) pageId: number,
    @Body() updateProjectPageDto: UpdateProjectPageDto,
  ): Promise<ProjectPageEntity> {
    return this.projectPageService.update(user, pageId, updateProjectPageDto);
  }

  @Delete(":id/page/:pageId")
  remove(
    @FirebaseUser() user: IFirebaseUser,
    @Param("pageId", ParseIntPipe) pageId: number,
  ): Promise<ProjectPageEntity> {
    return this.projectPageService.remove(user, pageId);
  }
}
