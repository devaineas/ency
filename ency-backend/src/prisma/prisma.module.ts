import { Module } from "@nestjs/common";
import { PrismaInterceptor } from "./prisma.interceptor";
import { PrismaService } from "./prisma.service";

@Module({
  providers: [PrismaService, PrismaInterceptor],
  exports: [PrismaService, PrismaInterceptor],
})
export class PrismaModule {}
