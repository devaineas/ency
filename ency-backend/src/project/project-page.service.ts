import {
  BadRequestException,
  Injectable,
  NotFoundException,
} from "@nestjs/common";
import { ProjectPage } from "@prisma/client";
import { IFirebaseUser } from "src/auth/interfaces/user.interface";
import { PrismaService } from "src/prisma/prisma.service";
import { CreateProjectPageDto } from "./dto/create-project-page.dto";
import { UpdateProjectPageDto } from "./dto/update-project-page.dto";
import {
  ProjectPageEntity,
  ProjectPageEntityNoData,
} from "./enitiies/projectPage.entity";

@Injectable()
export class ProjectPageService {
  constructor(private prisma: PrismaService) {}

  create(
    user: IFirebaseUser,
    createProjectPageDto: CreateProjectPageDto,
  ): Promise<ProjectPageEntity> {
    return this.prisma.projectPage.create({
      data: {
        name: "Untitled Page",
        data: "",
        project: {
          connect: {
            // Only the owner can add
            id_userUid: {
              id: createProjectPageDto.projectId,
              userUid: user.uid,
            },
          },
        },
      },
    });
  }

  findAll(
    user: IFirebaseUser,
    projectId: number,
  ): Promise<ProjectPageEntityNoData[]> {
    return this.prisma.projectPage.findMany({
      where: {
        project: {
          id: projectId,
          OR: [
            // Either thr project isPublic or is owned by the user
            {
              isPublic: true,
            },
            {
              userUid: user.uid,
            },
          ],
        },
      },
      select: {
        // Deselecting data
        id: true,
        name: true,
        projectId: true,
        data: false,
      },
    });
  }

  async findOne(
    user: IFirebaseUser,
    pageId: number,
  ): Promise<ProjectPageEntity> {
    const projectPage = await this.prisma.projectPage.findFirst({
      where: {
        id: pageId,
        OR: [
          // Either project isPublic or owned by the authorized user
          {
            project: {
              isPublic: true,
            },
          },
          {
            project: {
              userUid: user.uid,
            },
          },
        ],
      },
    });

    if (!projectPage) {
      throw new NotFoundException("Project with given id not found");
    }

    return projectPage;
  }

  async update(
    user: IFirebaseUser,
    pageId: number,
    updateProjectPageDto: UpdateProjectPageDto,
  ): Promise<ProjectPageEntity> {
    const projectPage = await this.prisma.projectPage.findUnique({
      where: {
        id: pageId,
      },
      select: {
        project: true,
      },
    });

    if (!projectPage)
      throw new NotFoundException("Page with given id does not exist");

    if (projectPage?.project.userUid === user.uid)
      throw new BadRequestException(
        "You are not authorized to perform this action",
      );

    return this.prisma.projectPage.update({
      where: {
        id: pageId,
      },
      data: {
        ...updateProjectPageDto,
      },
    });
  }

  async remove(
    user: IFirebaseUser,
    pageId: number,
  ): Promise<ProjectPageEntity> {
    const projectPage = await this.prisma.projectPage.findUnique({
      where: {
        id: pageId,
      },
      select: {
        project: true,
      },
    });

    if (!projectPage)
      throw new NotFoundException("Page with given id does not exist");

    if (projectPage?.project.userUid === user.uid)
      throw new BadRequestException(
        "You are not authorized to perform this action",
      );

    return this.prisma.projectPage.delete({
      where: {
        id: pageId,
      },
    });
  }
}
