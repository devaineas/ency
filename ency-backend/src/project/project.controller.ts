import {
  Controller,
  Get,
  Post,
  Body,
  Put,
  Param,
  Delete,
  ParseIntPipe,
  NotFoundException,
} from "@nestjs/common";
import { ProjectService } from "./project.service";
import { CreateProjectDto } from "./dto/create-project.dto";
import { UpdateProjectDto } from "./dto/update-project.dto";
import { FirebaseUser } from "../auth/decorators/user.decorator";
import { IFirebaseUser } from "../auth/interfaces/user.interface";
import { Public } from "../auth/auth.guard";
import { ProjectEntity } from "./enitiies/project.entity";

@Controller("projects")
export class ProjectController {
  constructor(private readonly projectService: ProjectService) {}

  @Post()
  create(
    @Body() createProjectDto: CreateProjectDto,
    @FirebaseUser() user?: IFirebaseUser,
  ): Promise<ProjectEntity> {
    return this.projectService.create(createProjectDto, user?.uid);
  }

  @Get()
  findAll(@FirebaseUser() user: IFirebaseUser): Promise<ProjectEntity[]> {
    return this.projectService.findAll(user, user.uid);
  }

  @Public()
  @Get("user/:uid")
  async findUsersProjects(
    @FirebaseUser() user: IFirebaseUser,
    @Param("uid") userUid: string,
  ): Promise<ProjectEntity[]> {
    return this.projectService.findAll(user, userUid);
  }

  @Public()
  @Get(":id")
  async findOne(
    @Param("id", ParseIntPipe) id: number,
    @FirebaseUser() user?: IFirebaseUser,
  ): Promise<ProjectEntity> {
    const project = await this.projectService.findOne(id);

    if (!project.isPublic && project.userUid !== user?.uid) {
      throw new NotFoundException(
        "A project with the given id not found (or is not public and not owned by you)",
      );
    }

    return project;
  }

  @Put(":id")
  update(
    @FirebaseUser() user: IFirebaseUser,
    @Param("id", ParseIntPipe) id: number,
    @Body() updateProjectDto: UpdateProjectDto,
  ): Promise<ProjectEntity> {
    return this.projectService.update(user.uid, id, updateProjectDto);
  }

  @Delete(":id")
  remove(
    @FirebaseUser() user: IFirebaseUser,
    @Param("id", ParseIntPipe) id: number,
  ) {
    return this.projectService.remove(user.uid, id);
  }
}
