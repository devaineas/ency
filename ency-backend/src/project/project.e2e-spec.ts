import { Test, TestingModule } from "@nestjs/testing";
import { INestApplication } from "@nestjs/common";
import Request from "supertest";
import { ProjectModule } from "./project.module";

describe("ProjectController (e2e)", () => {
  let app: INestApplication;

  beforeEach(async () => {
    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [ProjectModule],
    }).compile();

    app = moduleFixture.createNestApplication();
    await app.init();
  });

  it("/ (GET)", () => {
    return (
      Request(app.getHttpServer())
        // TODO
        .get("/")
        .expect(404)
    );
  });
});
