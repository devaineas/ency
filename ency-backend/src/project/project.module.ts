import { Module } from "@nestjs/common";
import { ProjectService } from "./project.service";
import { ProjectController } from "./project.controller";
import { PrismaModule } from "../prisma/prisma.module";
import { AuthModule } from "../auth/auth.module";
import { ProjectPageController } from "./project-page.controller";
import { ProjectPageService } from "./project-page.service";

@Module({
  imports: [PrismaModule, AuthModule],
  controllers: [ProjectController, ProjectPageController],
  providers: [ProjectService, ProjectPageService],
})
export class ProjectModule {}
