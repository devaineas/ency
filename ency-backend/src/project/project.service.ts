import {
  BadRequestException,
  Injectable,
  NotFoundException,
} from "@nestjs/common";
import { IFirebaseUser } from "src/auth/interfaces/user.interface";
import { PrismaService } from "../prisma/prisma.service";
import { CreateProjectDto } from "./dto/create-project.dto";
import { UpdateProjectDto } from "./dto/update-project.dto";
import { ProjectEntity } from "./enitiies/project.entity";

const FIRST_PAGE_TITLE = "Untitled page 1";

@Injectable()
export class ProjectService {
  constructor(private prisma: PrismaService) {}

  async create(
    createProjectDto: CreateProjectDto,
    userUid?: string,
  ): Promise<ProjectEntity> {
    if (!userUid) throw new BadRequestException();

    const project = await this.prisma.project.create({
      data: {
        ...createProjectDto,
        user: {
          connect: {
            uid: userUid,
          },
        },
        pages: {
          create: {
            name: FIRST_PAGE_TITLE,
            data: "",
          },
        },
      },
    });

    return {
      ...project,
      pageCount: 1,
    };
  }

  async findAll(
    user: IFirebaseUser,
    userUid: string,
  ): Promise<ProjectEntity[]> {
    const projects = await this.prisma.project.findMany({
      where: {
        userUid,
        OR: [
          // Either Project isPublic or is owned by authorized user
          {
            isPublic: true,
          },
          {
            userUid: user.uid,
          },
        ],
      },
    });

    // ISSUE: Change this to more performant one once this issue is resolved: https://github.com/prisma/prisma/issues/5079
    return Promise.all(
      projects.map(async (project) => {
        const pageCount = await this.findOnePageCount(project.id);
        return {
          ...project,
          pageCount,
        };
      }),
    );
  }

  async findOne(id: number): Promise<ProjectEntity> {
    const project = await this.prisma.project.findUnique({
      where: {
        id,
      },
    });

    if (!project)
      throw new NotFoundException(
        `A Project with the given id: ${id} not found`,
      );

    const pageCount = await this.findOnePageCount(project.id);

    return {
      ...project,
      pageCount,
    };
  }

  async findOnePageCount(projectId: number): Promise<number> {
    return this.prisma.projectPage.count({
      where: {
        projectId: projectId,
      },
    });
  }

  async update(
    userUid: string,
    id: number,
    updateProjectDto: UpdateProjectDto,
  ): Promise<ProjectEntity> {
    return await this.prisma.project.update({
      where: {
        id_userUid: {
          id: id,
          userUid: userUid,
        },
      },
      data: {
        ...updateProjectDto,
      },
    });
  }

  async remove(userUid: string, id: number): Promise<number> {
    // return this.prisma.project.delete({
    //   where: {
    //     id_userUid: {
    //       id: id,
    //       userUid: userUid,
    //     },
    //   },
    // });

    const res = await this.prisma.$executeRaw`
      DELETE FROM project WHERE "userUid" = ${userUid} AND id = ${id}
    `;

    if (res !== 1) {
      throw new NotFoundException("Project with given id not found");
    }

    return res;
  }
}
