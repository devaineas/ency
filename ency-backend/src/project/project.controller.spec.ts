import { Test, TestingModule } from "@nestjs/testing";
import { PrismaModule } from "../prisma/prisma.module";
import { ProjectController } from "./project.controller";
import { ProjectService } from "./project.service";
import { users, projects } from "../../prisma/mock-data.json";
import { Project } from "@prisma/client";
import { IFirebaseUser } from "../auth/interfaces/user.interface";
import { PrismaClientKnownRequestError } from "@prisma/client/runtime";
import { ProjectEntity } from "./enitiies/project.entity";
import { NotFoundException } from "@nestjs/common";

describe("ProjectController", () => {
  let controller: ProjectController;
  let module: TestingModule;

  beforeAll(async () => {
    module = await Test.createTestingModule({
      imports: [PrismaModule],
      controllers: [ProjectController],
      providers: [ProjectService],
    }).compile();

    controller = module.get<ProjectController>(ProjectController);
  });

  afterAll(async () => {
    module.close();
  });

  describe("Post() create", () => {
    it("Creating project with valid data should return valid response", async () => {
      const mockProjectData = projects[0];
      const { userUid, ...mockProject } = mockProjectData;
      const mockUser: IFirebaseUser = {
        uid: userUid,
      };

      const project: Project = await controller.create(mockProject, mockUser);
      expect(project).toMatchObject(mockProjectData);
    });

    it("Creating project with invalid user data should throw error", async () => {
      const mockProjectData = projects[0];
      const { userUid, ...mockProject } = mockProjectData;
      const mockUser: IFirebaseUser = {
        uid: "invalid-user-data",
      };

      await expect(
        controller.create(mockProject, mockUser),
      ).rejects.toBeInstanceOf(PrismaClientKnownRequestError);
    });
  });

  describe("Get() findAll", () => {
    it("Should return array of projects for valid user", async () => {
      const mockUserData = users[0];
      const user: IFirebaseUser = { uid: mockUserData.uid };

      const mockProjectsData = projects.filter(
        (project) => project.userUid === user.uid,
      );

      await Promise.all(
        mockProjectsData.map(async (project) => {
          const { userUid, ...mockProjectData } = project;
          await controller.create(mockProjectData, user);
        }),
      );

      const projectsRes: Project[] = await controller.findAll(user);
      expect(projectsRes).toMatchSnapshot("mock-projects-data");
    });
  });

  describe("Get(:id) findOne", () => {
    it("Should return valid Project data", async () => {
      // Create a mock project in database which belongs to a user
      const mockUserData = users[0];
      const user: IFirebaseUser = { uid: mockUserData.uid };

      const mockProject = projects.filter(
        (project) => project.userUid === user.uid,
      )[0];

      const { userUid, ...mockProjectData } = mockProject;
      const { id: newProjectId } = await controller.create(
        mockProjectData,
        user,
      );

      const projectRes: ProjectEntity = await controller.findOne(
        newProjectId,
        user,
      );

      expect(projectRes).toMatchSnapshot("mock-project-0-data");
    });

    it("Should return valid Project data with not user authenticated", async () => {
      // Create a mock project in database that is public
      const mockUserData = users[0];
      const user: IFirebaseUser = { uid: mockUserData.uid };

      const mockProject = projects.filter((project) => project.isPublic)[0];
      const { userUid, ...mockProjectData } = mockProject;
      const { id: newProjectId } = await controller.create(
        mockProjectData,
        user,
      );

      const projectRes: ProjectEntity = await controller.findOne(newProjectId);

      expect(projectRes).toMatchSnapshot("mock-project-isPublic-data");
    });

    it("Should throw NotFoundException for invalid auth and not public post", async () => {
      // Create a mock project in database which is not public and belongs to diff user
      const mockUserData = users[0];
      const user: IFirebaseUser = { uid: mockUserData.uid };

      // Get a project which is not public
      const mockProject = projects.filter((project) => !project.isPublic)[0];

      const { userUid, ...mockProjectData } = mockProject;
      const { id: newProjectId } = await controller.create(
        mockProjectData,
        user,
      );

      const invalidUser: IFirebaseUser = { uid: "invalid-user-uid" };

      await expect(
        controller.findOne(newProjectId, invalidUser),
      ).rejects.toBeInstanceOf(NotFoundException);
    });
  });

  describe("Put(:id) update", () => {
    let newProjectId: number;
    const mockUserData = users[0];
    const user: IFirebaseUser = { uid: mockUserData.uid };
    beforeAll(async () => {
      const mockProject = projects.filter((project) => project.isPublic)[0];
      const { userUid, ...mockProjectData } = mockProject;
      const newProjectData = await controller.create(mockProjectData, user);
      newProjectId = newProjectData.id;
    });

    it("Should return valid response on valid user and data", async () => {
      const projectRes = await controller.update(user, newProjectId, {
        description: "Some new description",
        tags: ["updated"],
        isPublic: false,
      });
      expect(projectRes).toMatchSnapshot("mock-project-updated");
    });

    it("Should throw error on invalid user", async () => {
      const invalidUser: IFirebaseUser = { uid: "invalid-user-id" };

      await expect(
        controller.update(invalidUser, newProjectId, {
          description: "some newer description",
        }),
      ).rejects.toBeInstanceOf(PrismaClientKnownRequestError);
    });
  });

  describe("Delete(:id) remove", () => {
    let newProjectId: number;
    const mockUserData = users[0];
    const user: IFirebaseUser = { uid: mockUserData.uid };
    beforeEach(async () => {
      const mockProject = projects.filter((project) => project.isPublic)[0];
      const { userUid, ...mockProjectData } = mockProject;
      const newProjectData = await controller.create(mockProjectData, user);
      newProjectId = newProjectData.id;
    });

    it("Should return valid response on valid user and data", async () => {
      const res = await controller.remove(user, newProjectId);

      expect(res).toEqual(1);

      await expect(
        controller.findOne(newProjectId, user),
      ).rejects.toBeInstanceOf(NotFoundException);
    });

    it("Should throw error on invalid user", async () => {
      const invalidUser: IFirebaseUser = { uid: "invalid-user-id" };
      await expect(
        controller.remove(invalidUser, newProjectId),
      ).rejects.toBeInstanceOf(NotFoundException);
    });
  });
});
