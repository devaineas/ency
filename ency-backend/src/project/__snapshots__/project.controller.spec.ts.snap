// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectController Get() findAll Should return array of projects for valid user: mock-projects-data 1`] = `
Array [
  Object {
    "color": "#ff7777",
    "description": "Mock description",
    "iconSrc": null,
    "id": 1,
    "isPublic": false,
    "pageCount": 1,
    "tags": Array [
      "tag",
    ],
    "title": "Hello world",
    "userUid": "abc",
  },
  Object {
    "color": "#ff7777",
    "description": "Mock description",
    "iconSrc": null,
    "id": 2,
    "isPublic": false,
    "pageCount": 1,
    "tags": Array [
      "tag",
    ],
    "title": "Hello world",
    "userUid": "abc",
  },
  Object {
    "color": "#ff7777",
    "description": "Mock description",
    "iconSrc": null,
    "id": 3,
    "isPublic": true,
    "pageCount": 1,
    "tags": Array [
      "hello",
    ],
    "title": "Hello world",
    "userUid": "abc",
  },
]
`;

exports[`ProjectController Get(:id) findOne Should return valid Project data with not user authenticated: mock-project-isPublic-data 1`] = `
Object {
  "color": "#ff7777",
  "description": "Mock description",
  "iconSrc": null,
  "id": 5,
  "isPublic": true,
  "pageCount": 1,
  "tags": Array [
    "hello",
  ],
  "title": "Hello world",
  "userUid": "abc",
}
`;

exports[`ProjectController Get(:id) findOne Should return valid Project data: mock-project-0-data 1`] = `
Object {
  "color": "#ff7777",
  "description": "Mock description",
  "iconSrc": null,
  "id": 4,
  "isPublic": false,
  "pageCount": 1,
  "tags": Array [
    "tag",
  ],
  "title": "Hello world",
  "userUid": "abc",
}
`;

exports[`ProjectController Put(:id) update Should return valid response on valid user and data: mock-project-updated 1`] = `
Object {
  "color": "#ffffff",
  "description": "Some new description",
  "iconSrc": null,
  "id": 7,
  "isPublic": false,
  "tags": Array [
    "updated",
  ],
  "title": "Hello world",
  "userUid": "abc",
}
`;
